parlai/agents/examples/transformer_variant.py

```diff
@@ -0,0 +1,206 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""
+Example code for specifying custom transformer variants.
+
+TransformerVariantAgent:
+- Minimal changes needed to:
+    - Swap out a high-level component (encoder)
+    - Swap out a low-level component (decoder->layer->self_attention)
+
+VerboseTransformerAgent:
+- Doesn't swap out anything
+- Fully specifies all components, for illustration
+
+ConfigurableTransformerAgent:
+- Swaps out components based on command line args
+"""
+from __future__ import annotations
+import torch
+from enum import Enum
+from typing import Dict, Optional, Tuple, Union
+
+from parlai.agents.transformer.modules import (
+    TransformerFFN,
+    MultiHeadAttention,
+    TransformerDecoder,
+    TransformerDecoderLayer,
+    TransformerEncoder,
+    TransformerEncoderLayer,
+    TransformerGeneratorModel,
+)
+from parlai.agents.transformer.transformer import TransformerGeneratorAgent
+from parlai.core.opt import Opt
+from parlai.core.params import ParlaiParser
+import parlai.utils.logging as logging
+
+
+###########################################
+# Transformer With Two Components Swapped #
+###########################################
+
+
+class TransformerVariantAgent(TransformerGeneratorAgent):
```

Comment:
Small unit tests for all of these please. Just something like


```python
opt = ParlaiParser(True, True).parse_kwargs(
    model='parlai.agents.examples.transformer_variant:TransformerVariantAgent',
    ffn_size=16,
    embeddingsize=16,
    dict_file='zoo:unittest/transformer_generator2/model.dict',
    skip_generation=True
)
model = create_agent(opt)
# send the model a single training example to ensure it can forward/backward
model.observe({'text': '1 2 3 4', 'labels': ['1 2 3 4'], 'episode_done': True})
model.act()
# send the model a single validation example
model.observe({'text': '1 2 3 4', 'eval_labels': ['1 2 3 4'], 'episode_done': True})
model.act()
```